ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.services;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import java.util.function.Supplier;
+import org.apache.kafka.clients.admin.Admin;
+
+public class ClientContextFactory {
+
+  private final MemoizedSupplier<Admin> memoizedAdminClientSupplier;
+  private final MemoizedSupplier<ConnectClient> memoizedConnectClientSupplier;
+  private final MemoizedSupplier<SchemaRegistryClient> memoizedSRClientSupplier;
+  private final MemoizedSupplier<SimpleKsqlClient> memoizedKsqlClientSupplier;
+
+
+  public ClientContextFactory(Supplier<Admin> adminClientSupplier,
+                              Supplier<ConnectClient> connectClientSupplier,
+                              Supplier<SchemaRegistryClient> srClientSupplier,
+                              Supplier<SimpleKsqlClient> ksqlClientSupplier
+  ) {
+    this.memoizedAdminClientSupplier = new MemoizedSupplier<Admin>(adminClientSupplier);
+    this.memoizedConnectClientSupplier = new MemoizedSupplier<ConnectClient>(connectClientSupplier);
+    this.memoizedSRClientSupplier = new MemoizedSupplier<SchemaRegistryClient>(srClientSupplier);
+    this.memoizedKsqlClientSupplier = new MemoizedSupplier<SimpleKsqlClient>(ksqlClientSupplier);
+  }
+
+  public MemoizedSupplier<Admin> getMemoizedAdminClientSupplier() {
+    return memoizedAdminClientSupplier;
+  }
+
+  public MemoizedSupplier<ConnectClient> getMemoizedConnectClientSupplier() {
+    return memoizedConnectClientSupplier;
+  }
+
+  public MemoizedSupplier<SchemaRegistryClient> getMemoizedSRClientSupplier() {
+    return memoizedSRClientSupplier;
+  }
+
+  public MemoizedSupplier<SimpleKsqlClient> getMemoizedKsqlClientSupplier() {
+    return memoizedKsqlClientSupplier;
+  }
+
+  public static class MemoizedSupplier<T> {
+
+    private Supplier<T> supplier;
+    private boolean isInitialized;
+    private T value;
+
+    public MemoizedSupplier(Supplier<T> supplierToMemoize) {
+      this.supplier = supplierToMemoize;
+      this.isInitialized = false;
+      this.value = null;
+    }
+
+    public T get() {
+      if (!isInitialized) {
+        value = supplier.get();
+        isInitialized = true;
+      }
+      return value;
+    }
+
+    public boolean isInitialized() {
+      return isInitialized;
```

Comment:
no need for `isInitialized` boolean.  Just do:

```suggestion
      return value != null;
```